File: app/src/main/java/com/emika/app/presentation/adapter/calendar/SubTaskAdapter.java
package com.emika.app.presentation.adapter.calendar;

import android.content.Context;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.CheckBox;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.emika.app.R;
import com.emika.app.data.network.pojo.subTask.SubTask;
import com.emika.app.presentation.adapter.profile.ItemTouchHelper.ItemTouchHelperAdapter;
import com.emika.app.presentation.viewmodel.calendar.CalendarViewModel;
import com.emika.app.presentation.viewmodel.calendar.TaskInfoViewModel;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class SubTaskAdapter extends RecyclerView.Adapter<SubTaskAdapter.ViewHolder> implements ItemTouchHelperAdapter {
    private static final String TAG = "SubTaskAdapter";

    private List<SubTask> taskList;
    private CalendarViewModel calendarViewModel;
    private TaskInfoViewModel taskInfoViewModel;
    public SubTaskAdapter(List<SubTask> taskList, CalendarViewModel calendarViewModel, TaskInfoViewModel taskInfoViewModel) {
        this.taskList = new ArrayList<>();
        for (int i = 0; i < taskList.size(); i++) {
            if (!taskList.get(i).getStatus().equals("deleted"))
                this.taskList.add(taskList.get(i));
        }
        this.calendarViewModel = calendarViewModel;
        this.taskInfoViewModel = taskInfoViewModel;
    }

    public List<SubTask> getTaskList() {
        return taskList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_subtask, parent, false);
        return new ViewHolder(view);
    }


    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        SubTask subTask = taskList.get(position);

        holder.body.setImeOptions(EditorInfo.IME_ACTION_DONE);
        holder.body.setRawInputType(InputType.TYPE_CLASS_TEXT);
        if (subTask.getNewTask()) {
            holder.body.requestFocus();
            subTask.setNewTask(false);
        } else holder.body.clearFocus();
        if (subTask.getName() != null)
            holder.body.setText(subTask.getName());

        if (subTask.getStatus().equals("done"))
            holder.checkBox.setChecked(true);
        else
            holder.checkBox.setChecked(false);

        holder.checkBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            if (taskInfoViewModel != null)
                if (isChecked) {
                    subTask.setStatus("done");
                    taskInfoViewModel.updateSubTask(subTask);
                } else {
                    subTask.setStatus("wip");
                    taskInfoViewModel.updateSubTask(subTask);
                }
            else if (calendarViewModel != null) {
                if (isChecked) {
                    subTask.setStatus("done");
                } else {
                    subTask.setStatus("wip");
                }
            }
        });

        holder.body.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
//                subTask.setName(holder.body.getText().toString());
            }

            @Override
            public void afterTextChanged(Editable s) {
                if (taskInfoViewModel != null) {
                    subTask.setName(holder.body.getText().toString());
                    taskInfoViewModel.updateSubTask(subTask);
                }
                if (calendarViewModel != null) {
                    subTask.setName(holder.body.getText().toString());
                }
            }
        });
    }


    @Override
    public int getItemCount() {
        return taskList.size();
    }

    public void addSubTask(SubTask subTask) {
        subTask.setNewTask(true);
        taskList.add(getItemCount(), subTask);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ConstraintLayout item;
        TextView body;
        CheckBox checkBox;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            item = itemView.findViewById(R.id.item_sub_task);
            body = itemView.findViewById(R.id.sub_task_item_body);
            checkBox = itemView.findViewById(R.id.sub_task_item_checkbox);
        }
    }

    @Override
    public void onItemMove(int fromPosition, int toPosition) {
        if (fromPosition < toPosition) {
            for (int i = fromPosition; i < toPosition; i++) {
                Collections.swap(taskList, i, i + 1);
            }
        } else {
            for (int i = fromPosition; i > toPosition; i--) {
                Collections.swap(taskList, i, i - 1);
            }
        }
        notifyItemMoved(fromPosition, toPosition);
    }

    @Override
    public void onItemDismiss(int position) {
        if (taskInfoViewModel != null) {
            taskList.get(position).setStatus("deleted");
            taskInfoViewModel.updateSubTask(taskList.get(position));
        }
        taskList.remove(position);
        notifyItemRemoved(position);
        notifyItemRangeChanged(position, getItemCount());
    }
}

File: app/src/main/java/com/emika/app/presentation/adapter/calendar/SubTaskAdapter.kt
package com.emika.app.presentation.adapter.calendar

import android.R.attr.password
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.util.Log
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import android.widget.Toast
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import com.emika.app.R
import com.emika.app.data.network.pojo.subTask.SubTask
import com.emika.app.presentation.adapter.ItemTouchHelper.ItemTouchHelperAdapter
import com.emika.app.presentation.viewmodel.calendar.CalendarViewModel
import com.emika.app.presentation.viewmodel.calendar.TaskInfoViewModel
import java.util.*


class SubTaskAdapter(taskList: List<SubTask>, calendarViewModel: CalendarViewModel?, taskInfoViewModel: TaskInfoViewModel?) : RecyclerView.Adapter<SubTaskAdapter.ViewHolder>(), ItemTouchHelperAdapter {
    var taskList: MutableList<SubTask?> = mutableListOf()
    private val calendarViewModel: CalendarViewModel?
    private val taskInfoViewModel: TaskInfoViewModel?

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_subtask, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val subTask = taskList[position]
        subTask!!.planOrder = position
        holder.body.imeOptions = EditorInfo.IME_ACTION_DONE
        holder.body.setRawInputType(InputType.TYPE_CLASS_TEXT)
        if (subTask.newTask) {
            holder.body.requestFocus()
            subTask.newTask = false
        } else holder.body.clearFocus()
        if (subTask.name != null) holder.body.text = subTask.name
        holder.checkBox.isChecked = subTask.status == "done"
//        if (position == itemCount - 1)
//        holder.body.setOnEditorActionListener { v, actionId, event ->
//            if (actionId == EditorInfo.IME_ACTION_DONE) {
//                if (position == taskList.size-1 ) {
//
//                    taskList.add(itemCount, SubTask().apply {newTask = true})
//                    notifyItemInserted(itemCount)
//                    notifyDataSetChanged()
//                    Log.d(TAG, taskList.size.toString())
//                }
//                true
//            } else false
//        }
        holder.checkBox.setOnCheckedChangeListener { buttonView: CompoundButton?, isChecked: Boolean ->
            if (taskInfoViewModel != null) if (isChecked) {
                subTask.status = "done"
                taskInfoViewModel.updateSubTask(subTask)
            } else {
                subTask.status = "wip"
                taskInfoViewModel.updateSubTask(subTask)
            } else if (calendarViewModel != null) {
                if (isChecked) {
                    subTask.status = "done"
                } else {
                    subTask.status = "wip"
                }
            }
        }

        holder.body.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
//                subTask.setName(holder.body.getText().toString());
            }

            override fun afterTextChanged(s: Editable) {
                if (taskInfoViewModel != null) {
                    subTask.name = holder.body.text.toString()
                    taskInfoViewModel.updateSubTask(subTask)
                }
                if (calendarViewModel != null) {
                    subTask.name = holder.body.text.toString()
                }
            }
        })
    }

    override fun getItemCount(): Int {
        return taskList.size
    }

    fun addSubTask(subTask: SubTask) {
        val subTasks = ArrayList<String>()
        subTask.status = "wip"
        subTask.name = ""
//        subTask.parentTaskId = task!!.id
//        subTask.assignee = assignee.id
//        subTask.planDate = task!!.planDate
//        subTask.deadlineDate = task!!.deadlineDate
//        subTask.companyId = task!!.companyId
//        subTask.projectId = task!!.projectId
//        subTask.sectionId = task!!.sectionId

//        for (i in taskList.indices) {
//            subTasks.add(taskList[i]!!.name)
//        }
//        taskInfoViewModel?.task?.subTaskList = subTasks
//        taskInfoViewModel?.updateTask(taskInfoViewModel?.task)
//        taskInfoViewModel?.addSubTask(subTask)
        subTask.duration = 60
        subTask.newTask = true
        taskList.add(itemCount, subTask)
        notifyItemInserted(itemCount)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var item: ConstraintLayout
        var body: TextView
        var checkBox: CheckBox

        init {
            item = itemView.findViewById(R.id.item_sub_task)
            body = itemView.findViewById(R.id.sub_task_item_body)
            checkBox = itemView.findViewById(R.id.sub_task_item_checkbox)
        }
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int) {
        val list: MutableList<String> = ArrayList()
        if (fromPosition < toPosition) {
            for (i in fromPosition until toPosition) {
                taskList[i]?.order = i + 1
                Collections.swap(taskList, i, i + 1)
            }
        } else {
            for (i in fromPosition downTo toPosition + 1) {
                taskList[i]?.order = i - 1
                Collections.swap(taskList, i, i - 1)

            }
        }
        this.taskList.sortWith(Comparator { lhs: SubTask?, rhs: SubTask? ->
            if (lhs!!.order === rhs!!.order) {
                return@Comparator 0
            } else if (lhs!!.order < rhs!!.order) {
                return@Comparator -1
            } else {
                return@Comparator 1
            }
        })
        for (subTask in taskList)
            list.add(subTask?.id!!)

        taskInfoViewModel?.updateTaskOrder(list)
            notifyItemMoved(fromPosition, toPosition)
    }

    override fun onItemDismiss(position: Int) {
        if (taskInfoViewModel != null) {
            taskList[position]!!.status = "deleted"
            taskInfoViewModel.updateSubTask(taskList[position])
        }
        taskList.removeAt(position)
        notifyItemRemoved(position)
        notifyItemRangeChanged(position, itemCount)
    }

    companion object {
        private const val TAG = "SubTaskAdapter"
    }

    init {
        this.taskList = ArrayList()
        for (i in taskList.indices) {
            if (taskList[i].status != "deleted") this.taskList.add(taskList[i])
        }
        this.taskList.sortWith(Comparator { lhs: SubTask?, rhs: SubTask? ->
            if (lhs!!.order === rhs!!.order) {
                return@Comparator 0
            } else if (lhs!!.order < rhs!!.order) {
                return@Comparator -1
            } else {
                return@Comparator 1
            }
        })
        this.calendarViewModel = calendarViewModel
        this.taskInfoViewModel = taskInfoViewModel
    }
}